processing/src/main/java/org/apache/druid/segment/filter/ExpressionFilter.java

```diff
@@ -55,6 +69,105 @@ public ExpressionFilter(final Supplier<Expr> expr, final FilterTuning filterTuni
+
+      // The hashcode and equals are to make SubclassesMustOverrideEqualsAndHashCodeTest stop complaining..
+      // DruidPredicateFactory doesn't really need equals or hashcode, in fact only the 'toString' method is called
+      // when testing equality of DimensionPredicateFilter, so it's the truly required method, but even that seems
+      // strange at best.
+      // Rather than tackle removing the annotation and reworking equality tests for now, will leave this to refactor
+      // as part of https://github.com/apache/druid/issues/8256 which suggests combining Filter and DimFilter
+      // implementations, which should clean up some of this mess.
+      @Override
+      public int hashCode()
+      {
+        return super.hashCode();
+      }
+
+      @Override
+      public boolean equals(Object obj)
+      {
+        return super.equals(obj);
+      }
+    };
+
+
+    final ExprType outputType = theExpr.getOutputType(factory);
+
+    if (outputType == null) {
+      // if an expression is vectorizable, but the output type is null, the result will be null (or the default
+      // value in default mode) because expression is either all null constants or missing columns
+
+      // in sql compatible mode, this means no matches ever, so just use the false matcher:
+      if (NullHandling.sqlCompatible()) {
+        return new FalseVectorMatcher(factory.getVectorSizeInspector());
+      }
+      // in default mode, just fallback to using a long matcher since nearly all boolean-ish expressions
+      // output a long value so it is probably a safe bet? idk, ending up here by using all null-ish things
+      // in default mode is dancing on the edge of insanity anyway...
+      return VectorValueMatcherColumnProcessorFactory.instance().makeLongProcessor(
+          ColumnCapabilitiesImpl.createSimpleNumericColumnCapabilities(ValueType.LONG),
+          ExpressionVectorSelectors.makeVectorValueSelector(factory, theExpr)
+      ).makeMatcher(predicateFactory);
+    }
+
+    switch (outputType) {
+      case LONG:
+        return VectorValueMatcherColumnProcessorFactory.instance().makeLongProcessor(
+            ColumnCapabilitiesImpl.createSimpleNumericColumnCapabilities(ValueType.LONG),
+            ExpressionVectorSelectors.makeVectorValueSelector(factory, theExpr)
+        ).makeMatcher(predicateFactory);
+      case DOUBLE:
+        return VectorValueMatcherColumnProcessorFactory.instance().makeDoubleProcessor(
+            ColumnCapabilitiesImpl.createSimpleNumericColumnCapabilities(ValueType.DOUBLE),
+            ExpressionVectorSelectors.makeVectorValueSelector(factory, theExpr)
+        ).makeMatcher(predicateFactory);
+      case STRING:
+        return VectorValueMatcherColumnProcessorFactory.instance().makeObjectProcessor(
+            // using 'numeric' capabilities creator so we are configured to NOT be dictionary encoded, etc
+            ColumnCapabilitiesImpl.createSimpleNumericColumnCapabilities(ValueType.STRING),
```

Comment:
`createSimpleNumericColumnCapabilities` is a weird method to call with type STRING. Is this a mistake? Or is the method just named poorly?

Reply:
it is strange, but on purpose for the reason the comment says since the numeric capabilities defaults are correct in this case. I will make sure all current users are numeric types, and if so then just make these capabilites from scratch because we should be counting on something geared towards numeric behavior in case those change in some unexpected way. Or, if it is used by other non-numbers then rename it to something more appropriate, like `createSimpleSingleValueColumnCapabilites` or similar

Reply:
I think one of those two options you mentioned would be good. IMO we shouldn't call a function named `createSimpleNumericColumnCapabilities` in this context, even if it happens to do the right thing, because it "looks wrong" and that trips people up that are trying to read through it.

Reply:
added a `createSimpleStringColumnCapabilities`